File: compiler/native.ts
import type { compile } from "./compile.js";
import type { SourceFile } from "./compile.js";
import type { FieldPlan } from "../server/plan.js";
import { parse, Kind } from "graphql";

const upper = (s: string) => s[0].toUpperCase() + s.slice(1);
export interface NativeField { name: string; coordinate: string; args: FieldPlan["args"]; type: string; optional: boolean; list: boolean; object: boolean }
export interface NativeModel { name: string; schemaType: string; language: string; proto: boolean; fields: NativeField[] }
export interface NativeConversion { from: NativeModel; to: NativeModel }

export function nativeModels(output: ReturnType<typeof compile>, sources: SourceFile[]) {
  const models: NativeModel[] = [];
  const declarations = sources.filter(s => s.native).map(s => {
    const definition = parse(s.body).definitions[0];
    if (definition.kind !== Kind.OPERATION_DEFINITION && definition.kind !== Kind.FRAGMENT_DEFINITION) throw new Error("Invalid native definition");
    return { ...s.native!, name: s.native!.declaration + "Data", document: s.body, operation: definition.name!.value };
  });
  const names = new Set<string>();
  for (const declaration of declarations) {
    const key = `${declaration.language}:${declaration.name}`;
    if (names.has(key)) throw new Error(`Duplicate native generated type ${declaration.name}`);
    names.add(key);
  }
  function addModel(name: string, schemaType: string, fields: FieldPlan[], language: string, proto: boolean): void {
    const mapped = fields.map(f => {
      const type = f.fields ? `${name}_${upper(f.responseName)}` : f.type;
      if (f.fields) addModel(type, f.type, f.fields, language, proto);
      return { name: f.responseName, coordinate: f.coordinate, args: f.args, type, optional: !f.required, list: f.list, object: !!f.fields };
    });
    models.push({ name, schemaType, fields: mapped, language, proto });
  }
  for (const fragment of output.fragmentPlans) if (fragment.native) {
    addModel(fragment.native.declaration + "Data", fragment.type, fragment.fields, fragment.native.language, false);
  }
  for (const language of ["swift", "kotlin"]) {
    for (const operation of output.operations.filter(o => declarations.some(d => d.language === language && d.operation === o.name))) {
      addModel(operation.name + "Response", "Query", operation.fields, language, true);
    }
  }
  function contains(from: NativeModel, to: NativeModel): boolean {
    return from.schemaType === to.schemaType && to.fields.every(field => {
      const f = from.fields.find(f => f.name === field.name);
      if (!f || f.coordinate !== field.coordinate || JSON.stringify(f.args) !== JSON.stringify(field.args) || f.optional !== field.optional || f.list !== field.list || f.object !== field.object) return false;
      return field.object ? contains(models.find(m => m.name === f.type && m.language === from.language)!, models.find(m => m.name === field.type && m.language === to.language)!) : f.type === field.type;
    });
  }
  const conversions: NativeConversion[] = [];
  for (const to of models.filter(m => !m.proto)) for (const from of models) {
    if (from.language === to.language && from.name !== to.name && contains(from, to)) conversions.push({ from, to });
  }
  return { models, conversions, declarations };
}

const swiftScalar: Record<string, string> = { ID: "String", String: "String", Int: "Int32", Float: "Double", Boolean: "Bool" };
const protoProperty = (s: string) => s.replace(/Url$/, "URL");

export function swiftModels(native: ReturnType<typeof nativeModels>): string {
  const output = ["// Generated from Swift GraphQL declarations.\n"];
  for (const declaration of native.declarations.filter(d => d.language === "swift" && d.kind === "query")) {
    output.push(`typealias ${declaration.name} = App_V1_${declaration.operation}Response`);
  }
  for (const model of native.models.filter(m => m.language === "swift" && !m.proto)) {
    output.push(`struct ${model.name}: Equatable {\n${model.fields.map(f => {
      let type = f.object ? f.type : swiftScalar[f.type];
      if (f.list) type = `[${type}]`;
      if (f.optional) type += "?";
      return `    let ${f.name}: ${type}`;
    }).join("\n")}\n}`);
  }
  for (const { from, to } of native.conversions.filter(c => c.from.language === "swift")) {
    const name = from.proto ? `App_V1_${from.name}` : from.name;
    output.push(`extension ${name} {\n    func as${to.name}() -> ${to.name} {\n        ${to.name}(\n${to.fields.map(f => {
      const property = from.proto ? protoProperty(f.name) : f.name;
      let value = `self.${property}`;
      if (f.object) {
        if (f.list) value += `.map { $0.as${f.type}() }`;
        else value += `${f.optional && !from.proto ? "?" : ""}.as${f.type}()`;
      }
      if (f.optional && from.proto) value = `self.has${upper(property)} ? ${value} : nil`;
      return `            ${f.name}: ${value}`;
    }).join(",\n")}\n        )\n    }\n}`);
  }
  return output.join("\n\n") + "\n";
}

File: internal/compiler/native.go
package compiler

import (
	"fmt"
	"reflect"
	"strings"

	"github.com/azrsh/fragment-colocation-with-grpc/internal/plan"
)

type NativeField struct {
	Name       string                  `json:"name"`
	Coordinate string                  `json:"coordinate"`
	Args       map[string]plan.Binding `json:"args"`
	Type       string                  `json:"type"`
	Optional   bool                    `json:"optional"`
	List       bool                    `json:"list"`
	Object     bool                    `json:"object"`
}
type NativeModel struct {
	Name       string        `json:"name"`
	SchemaType string        `json:"schemaType"`
	Fields     []NativeField `json:"fields"`
	Language   string        `json:"language"`
	Proto      bool          `json:"proto"`
}
type NativeConversion struct {
	From NativeModel `json:"from"`
	To   NativeModel `json:"to"`
}
type NativeDeclaration struct {
	NativeSource
	Name      string `json:"name"`
	Document  string `json:"document"`
	Operation string `json:"operation"`
}
type NativeOutput struct {
	Models       []NativeModel       `json:"models"`
	Conversions  []NativeConversion  `json:"conversions"`
	Declarations []NativeDeclaration `json:"declarations"`
}

func NativeModels(output *Output, sources []SourceFile) (*NativeOutput, error) {
	native := &NativeOutput{Models: []NativeModel{}, Conversions: []NativeConversion{}, Declarations: []NativeDeclaration{}}
	names := map[string]bool{}
	for _, s := range sources {
		if s.Native == nil {
			continue
		}
		d, err := parseSource(s)
		if err != nil {
			return nil, err
		}
		name := s.Native.Declaration + "Data"
		key := s.Native.Language + ":" + name
		if names[key] {
			return nil, fmt.Errorf("Duplicate native generated type %s", name)
		}
		names[key] = true
		operation := ""
		if len(d.Operations) > 0 {
			operation = d.Operations[0].Name
		} else {
			operation = d.Fragments[0].Name
		}
		native.Declarations = append(native.Declarations, NativeDeclaration{NativeSource: *s.Native, Name: name, Document: s.Body, Operation: operation})
	}
	var add func(string, string, []plan.Field, string, bool)
	add = func(name, schemaType string, fields []plan.Field, language string, proto bool) {
		mapped := []NativeField{}
		for _, f := range fields {
			typ := f.Type
			if f.Fields != nil {
				typ = name + "_" + upper(f.ResponseName)
				add(typ, f.Type, f.Fields, language, proto)
			}
			mapped = append(mapped, NativeField{Name: f.ResponseName, Coordinate: f.Coordinate, Args: f.Args, Type: typ, Optional: !f.Required, List: f.List, Object: f.Fields != nil})
		}
		native.Models = append(native.Models, NativeModel{Name: name, SchemaType: schemaType, Fields: mapped, Language: language, Proto: proto})
	}
	for _, f := range output.Fragments {
		if f.Native != nil {
			add(f.Native.Declaration+"Data", f.Type, f.Fields, f.Native.Language, false)
		}
	}
	for _, language := range []string{"swift", "kotlin"} {
		for _, o := range output.Operations {
			for _, d := range native.Declarations {
				if d.Language == language && d.Operation == o.Name {
					add(o.Name+"Response", "Query", o.Fields, language, true)
					break
				}
			}
		}
	}
	find := func(name, language string) NativeModel {
		for _, m := range native.Models {
			if m.Name == name && m.Language == language {
				return m
			}
		}
		return NativeModel{}
	}
	var contains func(NativeModel, NativeModel) bool
	contains = func(from, to NativeModel) bool {
		if from.SchemaType != to.SchemaType {
			return false
		}
		for _, field := range to.Fields {
			var f *NativeField
			for i := range from.Fields {
				if from.Fields[i].Name == field.Name {
					f = &from.Fields[i]
					break
				}
			}
			if f == nil || f.Coordinate != field.Coordinate || !reflect.DeepEqual(f.Args, field.Args) || f.Optional != field.Optional || f.List != field.List || f.Object != field.Object {
				return false
			}
			if field.Object {
				if !contains(find(f.Type, from.Language), find(field.Type, to.Language)) {
					return false
				}
			} else if f.Type != field.Type {
				return false
			}
		}
		return true
	}
	for _, to := range native.Models {
		if to.Proto {
			continue
		}
		for _, from := range native.Models {
			if from.Language == to.Language && from.Name != to.Name && contains(from, to) {
				native.Conversions = append(native.Conversions, NativeConversion{From: from, To: to})
			}
		}
	}
	return native, nil
}

func SwiftModels(native *NativeOutput) string {
	scalars := map[string]string{"ID": "String", "String": "String", "Int": "Int32", "Float": "Double", "Boolean": "Bool"}
	output := []string{"// Generated from Swift GraphQL declarations.\n"}
	for _, d := range native.Declarations {
		if d.Language == "swift" && d.Kind == "query" {
			output = append(output, "typealias "+d.Name+" = App_V1_"+d.Operation+"Response")
		}
	}
	for _, m := range native.Models {
		if m.Language != "swift" || m.Proto {
			continue
		}
		fields := []string{}
		for _, f := range m.Fields {
			typ := f.Type
			if !f.Object {
				typ = scalars[f.Type]
			}
			if f.List {
				typ = "[" + typ + "]"
			}
			if f.Optional {
				typ += "?"
			}
			fields = append(fields, "    let "+f.Name+": "+typ)
		}
		output = append(output, "struct "+m.Name+": Equatable {\n"+strings.Join(fields, "\n")+"\n}")
	}
	for _, c := range native.Conversions {
		from, to := c.From, c.To
		if from.Language != "swift" {
			continue
		}
		name := from.Name
		if from.Proto {
			name = "App_V1_" + name
		}
		fields := []string{}
		for _, f := range to.Fields {
			property := f.Name
			if from.Proto && strings.HasSuffix(property, "Url") {
				property = strings.TrimSuffix(property, "Url") + "URL"
			}
			value := "self." + property
			if f.Object {
				if f.List {
					value += ".map { $0.as" + f.Type + "() }"
				} else {
					if f.Optional && !from.Proto {
						value += "?"
					}
					value += ".as" + f.Type + "()"
				}
			}
			if f.Optional && from.Proto {
				value = "self.has" + upper(property) + " ? " + value + " : nil"
			}
			fields = append(fields, "            "+f.Name+": "+value)
		}
		output = append(output, "extension "+name+" {\n    func as"+to.Name+"() -> "+to.Name+" {\n        "+to.Name+"(\n"+strings.Join(fields, ",\n")+"\n        )\n    }\n}")
	}
	return strings.Join(output, "\n\n") + "\n"
}
